<h1 align = "center">
Data Audit Notebook
</h1>

## Imports and Setup

Before anything, let's need to import relevant and necessary libraries

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
import missingno as msno

from missingno import matrix
from missingno import heatmap

pd.set_option('display.max_columns',None)

import warnings
warnings.filterwarnings("ignore")

## Load Datasets

We can opt to upload the datsets by typing in the path to the datasets each time. However, since we have all the necessary datasets in one folder, we can load them in a cleaner way by defining the path first and then load each one using a shorter code each time we want to read them in. 

Alternatively, a cleaner version, and one we'll opt for here, is to load the datasets using a  helper function in pathlib.

In [2]:
from pathlib import Path

# Defining path to folder
data_path = Path(r"C:\Users\admin\Documents\Projects\maize-yield-optimization-rwanda\data\raw")

def load_csv (filename):
    return pd.read_csv(data_path/filename)

# Now use it to access the datasets

field_df = load_csv("Field_Level.csv")
plot_df = load_csv("Plot_Level.csv")
soil_df = load_csv("Soil_PL.csv")

Before we start auditing the dataframes that we have, we can look at a few things that will point us to the direction of what to look for:

We want to look at the shape of the respective datasets.

- The shape of a dataset helps us know the number of fields and records we have in the dataframe. 
- This way we can assess whether the dataframe is balanced (not having too many fields with too few records)
 

In [3]:
datasets = {
    "Field" : field_df,
    "Plot" : plot_df,
    "Soil" : soil_df   
}
for name,df in datasets.items():
    print(f"\n{name}")
    
    print("-"*30)
    
    print("Shape:",df.shape)


Field
------------------------------
Shape: (399, 19)

Plot
------------------------------
Shape: (1598, 27)

Soil
------------------------------
Shape: (736, 5)


The shape tells us more about the dimensions  of the data we have. For instance, we now know that there are 399 field records for 19 variables; 1598 plot observations for 27 variables; and 736 soil measurmeents across 5 variables. This now leads us to column audit.

## Column Audit


What we are intereste to see here is whethere there exists any suspicious variables. 

We'll be looking for variables that have been stored wrongly. For example:

- numeric variables stored as string
- datetime variables stored as string
- string variables stored as object etc.

In [6]:
display(field_df.info())

<class 'pandas.DataFrame'>
RangeIndex: 399 entries, 0 to 398
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   block_id               399 non-null    str    
 1   season                 399 non-null    str    
 2   district               399 non-null    str    
 3   cell                   399 non-null    str    
 4   agzone                 399 non-null    str    
 5   meg_env                399 non-null    str    
 6   lat                    399 non-null    float64
 7   long                   399 non-null    float64
 8   alt                    399 non-null    float64
 9   landscape_pos          399 non-null    str    
 10  slope_pos              242 non-null    str    
 11  fert_rank              399 non-null    str    
 12  soil_color             399 non-null    str    
 13  rockness               242 non-null    str    
 14  soil_depth             242 non-null    str    
 15  slope            

None

From the data info above, we can see that all fields from the field dataframe have been tagged correctly. 

Also, the majority of the fields have a high proportion of non-null values, meaning low percentage of missingness.

In [7]:
display(field_df.head())

,block_id,season,district,cell,agzone,meg_env,lat,long,alt,landscape_pos,slope_pos,fert_rank,soil_color,rockness,soil_depth,slope,trial_follower_gender,trialfield_home_min,cult_land
0,adfa31ad-95bf-453d-94f1-d163a97756ee,21A,Rwamagana,Akanzu,Eastern Ridges,Low-Mid-Dry,-2.021072,30.334707,1638.04,Hills,Backslope,medium,red_less,no_rocks,shallow,3,Male,4.0,20.0
1,d8807f25-e675-49ae-87a3-d228eca23c72,21A,Rwamagana,Akanzu,Eastern Ridges,Low-Mid-Dry,-2.021166,30.334817,1622.55,Hills,Backslope,medium,red_more,less_rocky,shallow,4,Male,5.0,19.0
2,262efa71-6ff4-4e18-b6f1-0c3d293504c0,21A,Rwamagana,Akanzu,Eastern Ridges,Low-Mid-Dry,-2.020386,30.338258,1610.46,Hills,Backslope,medium,red_less,no_rocks,deep,3,Male,15.0,18.0
3,95c016dd-908a-4f07-9e49-926e9a9f4f19,21A,Rwamagana,Akanzu,Eastern Ridges,Low-Mid-Dry,-2.020589,30.337583,1620.42,Hills,Backslope,medium,red_less,no_rocks,deep,3,Male,3.0,18.0
4,1d5c61a0-ca6b-4db8-87c3-5ca46c7cad6d,21A,Rwamagana,Akanzu,Eastern Ridges,Low-Mid-Dry,-2.018119,30.338658,1614.13,Hills,Backslope,medium,black_less,less_rocky,shallow,4,Female,5.0,10.0


A glimpse of the dataframe indicates that the values are correctly stored for each field

In [8]:
  
display(field_df.describe(include='numeric'))

TypeError: data type 'numeric' not understood

From the above output, we realize that in the plot dataset, planting_date, germ_count_date, urea_V6_date, urea_V10_date, silking_date, R3_date and harvest_date have been tagged as string variables instead of datetime. These highlight the priority areas for data cleaning. 

With this then, we already know we'll need to convert these variables to datetime for us to properly manipulate and analyze the data correctly.

## Missingness Analysis

While any research would be targetting to collect records for all variables, it is important to acknowledge that there may be some of the records that would fail to be collected

In this sectio, our interest is to not only assess whether there are missing variables, but to also ascertain how much of the records are missing for each variable.

We'll be assessing the percentage of missing variables for each variable with missingness. 

A good rule of thumb is to not only flag out variables with missing records, but to also ascertain what percentage is missing. Possibly even visualize the missingness.

In [ ]:
for name,df in datasets.items():
    
    missing = (df.isnull().sum()/len(df))*100
    missing = missing.sort_values(ascending=False)
    
    print("\n")
    print(name)
    
    missing_var = missing[missing > 0]
    display(missing_var)
    if len(missing_var) > 0:
        subset_df = df[missing_var.index]
        fig, ax = plt.subplots(
            nrows = 1,
            ncols = 2,
            figsize = (16,6)
        )
        msno.matrix(subset_df,ax = ax[0])
        ax[0].set_title(f"{name}: Missing matrix")
        
        msno.heatmap(subset_df,ax = ax[1])
        ax[1].set_title(f"{name}: Missing Heatmap")
        plt.tight_layout()
        plt.show()    

## Determining Missingness Mechanism

Here, we are interested in identifying whether the missingness earlier identified is:

- Missing Completely At Randon (MCAR),
- Missing At Random (MAR), or
- Missing Not At Random (MNAR)

Since our primary interest is to investigate missingness by a universal variable like Mega Environment, for the field variables, we'll simply investigae why missingness occurs for the missing variables.

In [ ]:
field_missing_vars = ['trialfield_home_min','slope_pos','soil_depth','rockness',
                'cult_land','trial_follower_gender']

In [ ]:
for var in field_missing_vars:
    
    # Creating missing indicator
    field_df[f'{var}_missing'] = field_df[var].isnull()
    
    print(f"\n{'='*50}")
    print(f"missingness Pattern for: {var}")
    print(f"{'='*50}")
    
    output = (
        field_df.groupby(f'{var}_missing')['meg_env']
        .value_counts(dropna = False)
    )
    print(output)

In [ ]:
# Create a grid
fig, axes = plt.subplots(
    nrows = 3,
    ncols = 2,
    figsize = (16,14)
)
axes = axes.flatten()
for i, var in enumerate(field_missing_vars):
    # Missingness indicator
    field_df[f'{var}_missing'] = field_df[var].isnull()
    
    sns.countplot(
        data = field_df,
        x = 'meg_env',
        hue = f'{var}_missing',
        ax = axes[i]
)
axes[i].set_title(f'{var}')
axes[i].set_xlabel('Mega Environment')
axes[i].set_ylabel('Count')
axes[i].tick_params(axis = 'x', rotation = 45)

# Remove individual legends
axes[i].legend_.remove()

# Create one share legend
handles, labels = axes[0].get_legend_handles_labels()
fig.legend(
    handles,
    ['Present','Missing'],
    title = 'Status',
    loc = 'upper right'
)

plt.suptitle(
    'Missingness Pattern Accross Variables by Mega Environment',
    fontsize = 16
)

plt.tight_layout()
plt.subplots_adjust(top = 0.94)
plt.show()

For the Plot dataframe, since we don't have the Mega Environment as part of the dataframe, we'll need to create the smallest possible merge of the datasets, albeit temporarily, to investigate why missingness occurs.

In [ ]:

plot_audit = plot_df.merge(
    field_df[
        ['block_id','meg_env']
    ],
    on = 'block_id',
    how = 'left'
)

In [ ]:
plot_missing_vars = ['urea_V10_date','prior_season_crop','silking_date',
                     'compost_quality','R3_date']

In [ ]:
for var in plot_missing_vars:
    
    # Creating missing indicator
    plot_audit[f'{var}_missing'] = plot_audit[var].isnull()
    
    print(f"\n{'='*50}")
    print(f"missingness Pattern for: {var}")
    print(f"{'='*50}")
    
    output = (
        plot_audit.groupby(f'{var}_missing')['meg_env']
        .value_counts(dropna = False)
    )
    print(output)

In [ ]:
# Create a grid
fig, axes = plt.subplots(
    nrows = 3,
    ncols = 2,
    figsize = (16,14)
)
axes = axes.flatten()
for i, var in enumerate(plot_missing_vars):
    # Missingness indicator
    plot_audit[f'{var}_missing'] = plot_audit[var].isnull()
    
    sns.countplot(
        data = plot_audit,
        x = 'meg_env',
        hue = f'{var}_missing',
        ax = axes[i]
)
axes[i].set_title(f'{var}')
axes[i].set_xlabel('Mega Environment')
axes[i].set_ylabel('Count')
axes[i].tick_params(axis = 'x', rotation = 45)

# Remove individual legends
axes[i].legend_.remove()

# Create one share legend
handles, labels = axes[0].get_legend_handles_labels()
fig.legend(
    handles,
    ['Present','Missing'],
    title = 'Status',
    loc = 'upper right'
)

plt.suptitle(
    'Missingness Pattern Accross Variables by Mega Environment',
    fontsize = 16
)

plt.tight_layout()
plt.subplots_adjust(top = 0.94)
plt.show()

Similarly, For the soil dataframe, since we only have the `plot_id` but no `block_id` or `meg_env` as variables, we'll have to perform a temporary merge so we can be able to investigate missingness by meg_env. 

For this, we'll need to merge the soil_df to the temporary plot_audit dataframe.

In [ ]:
soil_audit = soil_df.merge(
    plot_audit[
        ['plot_id','meg_env']
    ],
    on = 'plot_id',
    how = 'left'
)

In [ ]:
soil_missing_vars = ['tot_N','sand','org_C','pH']

In [ ]:
for var in soil_missing_vars:
    
    # Creating missing indicator
    soil_audit[f'{var}_missing'] = soil_audit[var].isnull()
    
    print(f"\n{'='*50}")
    print(f"missingness Pattern for: {var}")
    print(f"{'='*50}")
    
    output = (
        soil_audit.groupby(f'{var}_missing')['meg_env']
        .value_counts(dropna = False)
    )
    print(output)

In [ ]:
# Create a grid
fig, axes = plt.subplots(
    nrows = 2,
    ncols = 2,
    figsize = (16,10)
)
axes = axes.flatten()
for i, var in enumerate(soil_missing_vars):
    # Missingness indicator
    soil_audit[f'{var}_missing'] = soil_audit[var].isnull()
    
    sns.countplot(
        data = soil_audit,
        x = 'meg_env',
        hue = f'{var}_missing',
        ax = axes[i]
)
axes[i].set_title(f'{var}')
axes[i].set_xlabel('Mega Environment')
axes[i].set_ylabel('Count')
axes[i].tick_params(axis = 'x', rotation = 45)

# Remove individual legends
axes[i].legend_.remove()

# Create one share legend
handles, labels = axes[0].get_legend_handles_labels()
fig.legend(
    handles,
    ['Present','Missing'],
    title = 'Status',
    loc = 'upper right'
)

plt.suptitle(
    'Missingness Pattern Accross Variables by Mega Environment',
    fontsize = 16
)

plt.tight_layout()
plt.subplots_adjust(top = 0.94)
plt.show()

## Duplicate Analysis

In this section, we are interested in flagging complete duplicate records. This should give insight into whether, for example, there were repeated measurements. 

In [ ]:
for name, df in datasets.items():
    
    print(
        f"{name}:",
        df.duplicated().sum()
    )

## Identifier integrity check

In [ ]:
field_df['block_id'].nunique()

In [ ]:
plot_df['plot_id'].nunique()

In [ ]:
plot_counts = plot_audit.groupby(
    'block_id'
).size()

plot_counts.value_counts()

In [ ]:
soil_df['plot_id'].nunique()

## Merge Simulation

After investigating the potential issues with our datasets, we can now merge the datasets.

However, before merging, we need to validate join keys

In [ ]:
field_plot_m = plot_df.merge(
    field_df,
    on = 'block_id',
    how = 'left',
    indicator = True
)

# Check to validate join
field_plot_m['_merge'].value_counts()

We further check for row inflation before we can ascertain that the merged dataset would retain it's original integrity.

If the master dataframe retains its integrity, we move forward to merge with the soil_df

In [ ]:
print(
    "Before:",
    len(plot_df)
)

print(
    "After:",
    len(field_plot_m)
)

In [ ]:
field_plot_m.head()

## Outlier Analysis